# astyanax-genome-paper
Scripts and workflows used in Warren et al. (2021)

This repository contains scripts and workflow used to analyze deletions in the
paper: Warren et al. (2021) "A chromosome level genome of _Astyanax mexicanus_
surface fish for comparing population-specific genetic differences contributing
to trait evolution." _Nature Communications_ 12:1447.

## Generating structural variant calls
We generated structural variant calls with short reads for all of the libraries
listed in `library_list.txt`, using two short-read SV-callers:
[manta](https://github.com/Illumina/manta) and
[lumpy](https://github.com/arq5x/lumpy-sv). For lumpy, we used the
[smoove](https://github.com/brentp/smoove) pipeline as recommended.
[nextflow](https://www.nextflow.io/) workflows for both are in
`1-generate-sv-calls`. Please refer to the webpages for these software packages
for requirements and installation instructions.

## Filtering structural variant calls
We filtered the structural variant calls generated by manta and lumpy to
include only deletions within the size range (500,100000). A shell script for
this is in `2-filter-sv-calls`.

## Analyzing and counting structural variant calls
We analyzed the deletion calls by counting the numbers of deletions per sample,
merging the lumpy and manta results, and annotating the deletions based on
whether they contain intronic, regulatory, or coding sequence, as described in
the manuscript. Python scripts for counting, merging, and annotating are in
`3-analyze-sv-calls`, as well as a bash script containing the commands we used
to run these programs.

## More information
Please consult the paper for more information about the methods and results,
and direct questions to the corresponding authors.
